### PR TITLE
Auto-generate the `hostcalls` module of `wasi-common`

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit updates the `wasi-common` crate to remove a souce of duplication from the `*.witx` files to what's listed in the source. Ideally we're now getting to the point where adding a WASI API means you update the `*.witx` file, then you write an implementation, then you're done. No further shims you need to write!

Some changes in this PR are:

* This is built on https://github.com/bytecodealliance/wasmtime/pull/845
* The `hostcalls` module, both for the current snapshot and old snapshot, are entirely generated from the `*.witx` files.
* The code generated should almost match what we had before, except that `&mut WasiCtx` is required everywhere instead of sometimes using `&WasiCtx` and sometimes `&mut WasiCtx`. Note that the handwritten bindings continue to use a mixture.
* The `wasi-common-cbindgen` crate has been removed since its functionality is easy enough to fold into the existing macro.

The general goal here is to continue to reduce duplication between all the WASI API definitions, ensuring that there's one source of truth for the signature and once source of truth for the implementation, and everything in the middle is auto-generated as necessary.

### CODE REVIEW

File: crates/wasi-common/src/error.rs [hunk 1]
```diff
@@ -242,6 +248,280 @@ impl Error {
...
+            return std::io::Error::new(std::io::ErrorKind::Other, error_str(errno))
+        }
+        _ => panic!("unrecognized WASI errno value"),
+    } as i32;
+
+    std::io::Error::from_raw_os_error(raw_os_error)
+}
+
+#[cfg(windows)]
+fn error_str(errno: wasi::__wasi_errno_t) -> &'static str {
```
Comment (kubkon) [inline crates/wasi-common/src/error.rs]: Hmm, I guess I must be missing something here but why do we need that here at all? Shouldn't we be able to generate this (as was before as far as I remember) directly from `*.witx`?

File: crates/wasi-common/src/error.rs [hunk 2]
```diff
@@ -242,6 +248,280 @@ impl Error {
...
 pub(crate) trait FromRawOsError {
     fn from_raw_os_error(code: i32) -> Self;
 }
+
+/// Translate a WASI errno code into an `io::Result<()>`.
+///
+/// TODO: Would it be better to have our own version of `io::Error` (and
+/// `io::Result`), rather than trying to shoehorn WASI errors into the
+/// libstd version?
+fn wasi_errno_to_io_error(errno: wasi::__wasi_errno_t) -> std::io::Error {
```
Comment (kubkon) [inline crates/wasi-common/src/error.rs]: This looks the wrong way round for me. As a temp solution, this is fine, but in the long run either, as you suggested, we should have our own version of `io::Error` or flatten all errors into WASI errno.

Actually, why do we need to convert the errno into `io::Error` in the first place? As far as I can remember, we've only ever done it the opposite way in this crate.

Comment (kubkon) [reply crates/wasi-common/src/error.rs]: OK, I've only now realised this is a simple copy-paste from `fs/error.rs`. Nvm then :-)

**Comment by kubkon** [comment]
> LGTM, just a couple of questions. Also, just an FYI so that we don't duplicate the effort that @pchickey and myself are currently working on an upgraded version of `wig`, the `wiggle` crate which should help us use the polyfill functionality of the `witx` crate to generate `wasi_unstable` and any future snapshots in a way that almost completely removes any unnecessary duplication in hostcalls implementation, etc. It should also help us auto-generate the `memory.rs` module.

Oh, and the link to [`wiggle`](https://github.com/kubkon/wiggle).

File: crates/wasi-common/src/error.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/wasi-common/src/error.rs]: Ah ok, if you'd like to dig in here (which sounds good to me!) I'll go ahead and keep refactoring this, and with some tweaks of error types this was easily removable.

File: crates/wasi-common/src/error.rs (hunk 2 above)
Comment (alexcrichton) [reply crates/wasi-common/src/error.rs]: It's true yeah it was a copy/paste, but I went ahead and refactored a bit more and was able to delete this function entirely.

**Comment by alexcrichton** [comment]
> Also, just an FYI so that we don't duplicate the effort that @pchickey and myself are currently working on 

Nice! I'm happy to go whichever way here, was just eager to help cut down on more boilerplate, and I also needed this for a refactoring of the `wasmtime-wasi` crate to use the `wasmtime` crate

**Pull request merged**